File: backend/app/schemas/book.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.book import BookFormat


class BookCreate(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    author: str = Field(min_length=1, max_length=255)
    series: str | None = None
    book_no: int | None = None
    genre: str | None = None
    cover_color: str = "#3F5D4E"
    cover_url: str | None = None
    published: str | None = None
    pages: int | None = None
    format: BookFormat = BookFormat.physical

    purchased: bool = False
    read: bool = False
    start_date: date | None = None
    end_date: date | None = None

    rating_cover: int | None = Field(default=None, ge=0, le=5)
    rating_writing: int | None = Field(default=None, ge=0, le=5)
    rating_plot: int | None = Field(default=None, ge=0, le=5)
    rating_characters: int | None = Field(default=None, ge=0, le=5)
    enjoyed: bool | None = None
    read_again: bool | None = None
    liked_most: str | None = None
    liked_least: str | None = None
    feel: str | None = None
    trope: str | None = None
    final_review: str | None = None
    favorite_characters: list[str] = Field(default_factory=list)
    notable_scenes: list[str] = Field(default_factory=list)
    quotes: list[str] = Field(default_factory=list)


class BookUpdate(BaseModel):
    title: str | None = None
    author: str | None = None
    series: str | None = None
    book_no: int | None = None
    genre: str | None = None
    cover_color: str | None = None
    cover_url: str | None = None
    published: str | None = None
    pages: int | None = None
    format: BookFormat | None = None

    purchased: bool | None = None
    read: bool | None = None
    times_read: int | None = None
    start_date: date | None = None
    end_date: date | None = None

    rating_cover: int | None = Field(default=None, ge=0, le=5)
    rating_writing: int | None = Field(default=None, ge=0, le=5)
    rating_plot: int | None = Field(default=None, ge=0, le=5)
    rating_characters: int | None = Field(default=None, ge=0, le=5)
    enjoyed: bool | None = None
    read_again: bool | None = None
    liked_most: str | None = None
    liked_least: str | None = None
    feel: str | None = None
    trope: str | None = None
    final_review: str | None = None
    favorite_characters: list[str] | None = None
    notable_scenes: list[str] | None = None
    quotes: list[str] | None = None


class BookOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    author: str
    series: str | None
    book_no: int | None
    genre: str | None
    cover_color: str
    cover_url: str | None
    published: str | None
    pages: int | None
    format: BookFormat

    purchased: bool
    read: bool
    times_read: int
    start_date: date | None
    end_date: date | None

    rating: float | None
    rating_cover: int | None
    rating_writing: int | None
    rating_plot: int | None
    rating_characters: int | None
    enjoyed: bool | None
    read_again: bool | None
    liked_most: str | None
    liked_least: str | None
    feel: str | None
    trope: str | None
    final_review: str | None
    favorite_characters: list[str]
    notable_scenes: list[str]
    quotes: list[str]

    created_at: datetime
    updated_at: datetime


class BookSearchResult(BaseModel):
    title: str
    author: str
    cover_url: str | None = None
    pages: int | None = None
    published: str | None = None
    genre: str | None = None

File: app/lib/models/book.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class Book {
  final String id;
  final String title;
  final String author;
  final String? series;
  final int? bookNo;
  final String? genre;
  final String coverColor;
  final String? coverUrl;
  final String? published;
  final int? pages;
  final String format;

  final bool purchased;
  final bool read;
  final int timesRead;
  final DateTime? startDate;
  final DateTime? endDate;

  final double? rating;
  final int? ratingCover;
  final int? ratingWriting;
  final int? ratingPlot;
  final int? ratingCharacters;
  final bool? enjoyed;
  final bool? readAgain;
  final String? likedMost;
  final String? likedLeast;
  final String? feel;
  final String? trope;
  final String? finalReview;
  final List<String> favoriteCharacters;
  final List<String> notableScenes;
  final List<String> quotes;

  Book({
    required this.id,
    required this.title,
    required this.author,
    this.series,
    this.bookNo,
    this.genre,
    required this.coverColor,
    this.coverUrl,
    this.published,
    this.pages,
    required this.format,
    required this.purchased,
    required this.read,
    required this.timesRead,
    this.startDate,
    this.endDate,
    this.rating,
    this.ratingCover,
    this.ratingWriting,
    this.ratingPlot,
    this.ratingCharacters,
    this.enjoyed,
    this.readAgain,
    this.likedMost,
    this.likedLeast,
    this.feel,
    this.trope,
    this.finalReview,
    required this.favoriteCharacters,
    required this.notableScenes,
    required this.quotes,
  });

  String get initials {
    final words = title.split(RegExp(r'\s+')).where((w) => w.isNotEmpty).toList();
    return words.take(3).map((w) => w[0].toUpperCase()).join();
  }

  bool get isReading => purchased && !read && startDate != null;

  static DateTime? _parseDate(dynamic v) => v == null ? null : (v as Timestamp).toDate();
  static List<String> _parseStrings(dynamic v) => (v as List?)?.map((e) => e as String).toList() ?? const [];

  factory Book.fromFirestore(DocumentSnapshot<Map<String, dynamic>> doc) {
    final data = doc.data() ?? const {};
    return Book(
      id: doc.id,
      title: data['title'] as String? ?? '',
      author: data['author'] as String? ?? '',
      series: data['series'] as String?,
      bookNo: data['bookNo'] as int?,
      genre: data['genre'] as String?,
      coverColor: data['coverColor'] as String? ?? '#3F5D4E',
      coverUrl: data['coverUrl'] as String?,
      published: data['published'] as String?,
      pages: data['pages'] as int?,
      format: data['format'] as String? ?? 'physical',
      purchased: data['purchased'] as bool? ?? false,
      read: data['read'] as bool? ?? false,
      timesRead: data['timesRead'] as int? ?? 0,
      startDate: _parseDate(data['startDate']),
      endDate: _parseDate(data['endDate']),
      rating: (data['rating'] as num?)?.toDouble(),
      ratingCover: data['ratingCover'] as int?,
      ratingWriting: data['ratingWriting'] as int?,
      ratingPlot: data['ratingPlot'] as int?,
      ratingCharacters: data['ratingCharacters'] as int?,
      enjoyed: data['enjoyed'] as bool?,
      readAgain: data['readAgain'] as bool?,
      likedMost: data['likedMost'] as String?,
      likedLeast: data['likedLeast'] as String?,
      feel: data['feel'] as String?,
      trope: data['trope'] as String?,
      finalReview: data['finalReview'] as String?,
      favoriteCharacters: _parseStrings(data['favoriteCharacters']),
      notableScenes: _parseStrings(data['notableScenes']),
      quotes: _parseStrings(data['quotes']),
    );
  }
}
